### ISSUE TITLE
Integrity error - Cannot add or update a child row: a foreign key constraint fails

### ISSUE THREAD

**Issue by bpcurse** [pretix/pretix#1280]
There were many normal booking events today and Pretix seems to run smoothly but this was found in the logs as a single occurrence. Could this be a bug?
Running Pretix 2.6.0, no-docker, mariadb 10.3.13, php 7.3.3, nginx reverse proxy

```
Mai 10 14:33:13 some-server celery[14928]: [2019-05-10 12:33:13,741: INFO/MainProcess] Received task: pretix.base.services.cart.clear_cart[db1e3c44-0341-4e8c-a969-9b581f841e51]
Mai 10 14:33:14 some-server celery[14928]: [2019-05-10 12:33:14,060: ERROR/ForkPoolWorker-5] Task pretix.base.services.cart.clear_cart[db1e3c44-0341-4e8c-a969-9b581f841e51] raised unexpected: IntegrityError(1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')
Mai 10 14:33:14 some-server celery[14928]: Traceback (most recent call last):
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Mai 10 14:33:14 some-server celery[14928]:     return self.cursor.execute(sql, params)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
Mai 10 14:33:14 some-server celery[14928]:     return self.cursor.execute(query, args)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 206, in execute
Mai 10 14:33:14 some-server celery[14928]:     res = self._query(query)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 312, in _query
Mai 10 14:33:14 some-server celery[14928]:     db.query(q)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/connections.py", line 224, in query
Mai 10 14:33:14 some-server celery[14928]:     _mysql.connection.query(self, query)
Mai 10 14:33:14 some-server celery[14928]: MySQLdb._exceptions.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')
Mai 10 14:33:14 some-server celery[14928]: The above exception was the direct cause of the following exception:
Mai 10 14:33:14 some-server celery[14928]: Traceback (most recent call last):
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/celery/app/trace.py", line 375, in trace_task
Mai 10 14:33:14 some-server celery[14928]:     R = retval = fun(*args, **kwargs)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/tasks.py", line 39, in __call__
Mai 10 14:33:14 some-server celery[14928]:     ret = super().__call__(*args, **kwargs)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/celery/app/trace.py", line 632, in __protected_call__
Mai 10 14:33:14 some-server celery[14928]:     return self.run(*args, **kwargs)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/cart.py", line 915, in clear_cart
Mai 10 14:33:14 some-server celery[14928]:     cm.commit()
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/cart.py", line 784, in commit
Mai 10 14:33:14 some-server celery[14928]:     err = self._perform_operations() or err
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/cart.py", line 758, in _perform_operations
Mai 10 14:33:14 some-server celery[14928]:     op.position.save()
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/models/orders.py", line 1023, in save
Mai 10 14:33:14 some-server celery[14928]:     super().save(*args, **kwargs)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 718, in save
Mai 10 14:33:14 some-server celery[14928]:     force_update=force_update, update_fields=update_fields)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 748, in save_base
Mai 10 14:33:14 some-server celery[14928]:     updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 831, in _save_table
Mai 10 14:33:14 some-server celery[14928]:     result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 869, in _do_insert
Mai 10 14:33:14 some-server celery[14928]:     using=using, raw=raw)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
Mai 10 14:33:14 some-server celery[14928]:     return getattr(self.get_queryset(), name)(*args, **kwargs)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/query.py", line 1136, in _insert
Mai 10 14:33:14 some-server celery[14928]:     return query.get_compiler(using=using).execute_sql(return_id)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1289, in execute_sql
Mai 10 14:33:14 some-server celery[14928]:     cursor.execute(sql, params)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
Mai 10 14:33:14 some-server celery[14928]:     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
Mai 10 14:33:14 some-server gunicorn[14915]: ERROR 2019-05-10 12:33:14,062 pretix.base.tasks tasks Unexpected exception: IntegrityError("(1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')",)
Mai 10 14:33:14 some-server gunicorn[14915]: ERROR 2019-05-10 12:33:14,064 pretix.base.tasks tasks Unexpected exception: IntegrityError("(1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')",)
Mai 10 14:33:14 some-server celery[14928]:     return executor(sql, params, many, context)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Mai 10 14:33:14 some-server celery[14928]:     return self.cursor.execute(sql, params)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
Mai 10 14:33:14 some-server celery[14928]:     raise dj_exc_value.with_traceback(traceback) from exc_value
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Mai 10 14:33:14 some-server celery[14928]:     return self.cursor.execute(sql, params)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
Mai 10 14:33:14 some-server celery[14928]:     return self.cursor.execute(query, args)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 206, in execute
Mai 10 14:33:14 some-server celery[14928]:     res = self._query(query)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 312, in _query
Mai 10 14:33:14 some-server celery[14928]:     db.query(q)
Mai 10 14:33:14 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/connections.py", line 224, in query
Mai 10 14:33:14 some-server celery[14928]:     _mysql.connection.query(self, query)
Mai 10 14:33:14 some-server celery[14928]: django.db.utils.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')
Mai 10 14:33:22 some-server celery[14928]: [2019-05-10 12:33:22,540: INFO/MainProcess] Received task: pretix.base.services.cart.clear_cart[5a9e9c1d-b120-4e62-b57a-091e30b3ab2b]
Mai 10 14:33:22 some-server celery[14928]: [2019-05-10 12:33:22,843: ERROR/ForkPoolWorker-7] Task pretix.base.services.cart.clear_cart[5a9e9c1d-b120-4e62-b57a-091e30b3ab2b] raised unexpected: IntegrityError(1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')
Mai 10 14:33:22 some-server celery[14928]: Traceback (most recent call last):
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Mai 10 14:33:22 some-server celery[14928]:     return self.cursor.execute(sql, params)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
Mai 10 14:33:22 some-server celery[14928]:     return self.cursor.execute(query, args)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 206, in execute
Mai 10 14:33:22 some-server celery[14928]:     res = self._query(query)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 312, in _query
Mai 10 14:33:22 some-server celery[14928]:     db.query(q)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/connections.py", line 224, in query
Mai 10 14:33:22 some-server celery[14928]:     _mysql.connection.query(self, query)
Mai 10 14:33:22 some-server celery[14928]: MySQLdb._exceptions.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')
Mai 10 14:33:22 some-server celery[14928]: The above exception was the direct cause of the following exception:
Mai 10 14:33:22 some-server celery[14928]: Traceback (most recent call last):
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/celery/app/trace.py", line 375, in trace_task
Mai 10 14:33:22 some-server celery[14928]:     R = retval = fun(*args, **kwargs)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/tasks.py", line 39, in __call__
Mai 10 14:33:22 some-server celery[14928]:     ret = super().__call__(*args, **kwargs)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/celery/app/trace.py", line 632, in __protected_call__
Mai 10 14:33:22 some-server celery[14928]:     return self.run(*args, **kwargs)
Mai 10 14:33:22 some-server gunicorn[14915]: ERROR 2019-05-10 12:33:22,844 pretix.base.tasks tasks Unexpected exception: IntegrityError("(1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')",)
Mai 10 14:33:22 some-server gunicorn[14915]: ERROR 2019-05-10 12:33:22,846 pretix.base.tasks tasks Unexpected exception: IntegrityError("(1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')",)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/cart.py", line 915, in clear_cart
Mai 10 14:33:22 some-server celery[14928]:     cm.commit()
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/cart.py", line 784, in commit
Mai 10 14:33:22 some-server celery[14928]:     err = self._perform_operations() or err
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/services/cart.py", line 758, in _perform_operations
Mai 10 14:33:22 some-server celery[14928]:     op.position.save()
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/pretix/base/models/orders.py", line 1023, in save
Mai 10 14:33:22 some-server celery[14928]:     super().save(*args, **kwargs)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 718, in save
Mai 10 14:33:22 some-server celery[14928]:     force_update=force_update, update_fields=update_fields)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 748, in save_base
Mai 10 14:33:22 some-server celery[14928]:     updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 831, in _save_table
Mai 10 14:33:22 some-server celery[14928]:     result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/base.py", line 869, in _do_insert
Mai 10 14:33:22 some-server celery[14928]:     using=using, raw=raw)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
Mai 10 14:33:22 some-server celery[14928]:     return getattr(self.get_queryset(), name)(*args, **kwargs)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/query.py", line 1136, in _insert
Mai 10 14:33:22 some-server celery[14928]:     return query.get_compiler(using=using).execute_sql(return_id)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1289, in execute_sql
Mai 10 14:33:22 some-server celery[14928]:     cursor.execute(sql, params)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
Mai 10 14:33:22 some-server celery[14928]:     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
Mai 10 14:33:22 some-server celery[14928]:     return executor(sql, params, many, context)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Mai 10 14:33:22 some-server celery[14928]:     return self.cursor.execute(sql, params)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
Mai 10 14:33:22 some-server celery[14928]:     raise dj_exc_value.with_traceback(traceback) from exc_value
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Mai 10 14:33:22 some-server celery[14928]:     return self.cursor.execute(sql, params)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/django/db/backends/mysql/base.py", line 71, in execute
Mai 10 14:33:22 some-server celery[14928]:     return self.cursor.execute(query, args)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 206, in execute
Mai 10 14:33:22 some-server celery[14928]:     res = self._query(query)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/cursors.py", line 312, in _query
Mai 10 14:33:22 some-server celery[14928]:     db.query(q)
Mai 10 14:33:22 some-server celery[14928]:   File "/var/pretix/venv/lib/python3.6/site-packages/MySQLdb/connections.py", line 224, in query
Mai 10 14:33:22 some-server celery[14928]:     _mysql.connection.query(self, query)
Mai 10 14:33:22 some-server celery[14928]: django.db.utils.IntegrityError: (1452, 'Cannot add or update a child row: a foreign key constraint fails (`pretix2`.`pretixbase_cartposition`, CONSTRAINT `pretixbase_cartposit_addon_to_id_d9ab2dee_fk_pretixbas` FOREIGN KEY (`addon_to_id`) REFERENCES `pretixbase_cartposition` (`id`))')
```

**Comment by raphaelm** [comment]
Hmm, might be, looks like removing things from cart didn't work for a user. We resolved a number of these bugs in the last months, but I think 2.6 should already be safe, not 100% sure, though. Do you use add-on products?

**Comment by bpcurse** [comment]
Some additional info from the nginx access log:
```
xxx.xxx.xxx.xxx - - [10/May/2019:14:33:08 +0200] "GET /xxxxx/xxxxxxm2019/ HTTP/2.0" 200 53664 "-" "Mozilla/5.0 (Linux; Android 8.0.0; SM-G935F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36" "-"
xxx.xxx.xxx.xxx - - [10/May/2019:14:33:14 +0200] "POST /xxxxx/xxxxxxm2019/cart/clear HTTP/2.0" 200 304 "https://pretix.xxxxx.de/" "Mozilla/5.0 (Linux; Android 8.0.0; SM-G935F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36" "-"
xxx.xxx.xxx.xxx - - [10/May/2019:14:33:15 +0200] "GET /xxxxx/xxxxxm2019/?require_cookie=true HTTP/2.0" 200 53808 "https://pretix.xxxxx.de/" "Mozilla/5.0 (Linux; Android 8.0.0; SM-G935F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36" "-"
xxx.xxx.xxx.xxx - - [10/May/2019:14:33:22 +0200] "POST /xxxxx/xxxxxm2019/cart/clear HTTP/2.0" 200 304 "https://pretix.xxxxx.de/" "Mozilla/5.0 (Linux; Android 8.0.0; SM-G935F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36" "-"
xxx.xxx.xxx.xxx - - [10/May/2019:14:33:23 +0200] "GET /xxxxx/xxxxxm2019/?require_cookie=true HTTP/2.0" 200 53808 "https://pretix.xxxxx.de/" "Mozilla/5.0 (Linux; Android 8.0.0; SM-G935F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.90 Mobile Safari/537.36" "-"
```

There have been no errors reported in nginx error log.
Regarding the add-on products, I will have to ask the event organizers.

@raphaelm Thank you for correcting the messy log display in my initial post :)

**Comment by David100mark** [comment]
Yes, we do use add-on products for this event.

**Comment by raphaelm** [comment]
On second thought, I'm 99% sure that this has been fixed with pretix 2.7.0 in https://github.com/pretix/pretix/commit/2810e2a

If not, please reopen after upgrade :)

**Comment by tutorialcreation** [comment]
Great
